TypeCobol.Analysis/Graph/ControlFlowGraph.cs

```diff
@@ -0,0 +1,170 @@
+        /// </summary>
+        public Dictionary<N, BasicBlock<N,D>> BlockFor
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The list of all Successor edges. The successor list for each basic block is a sublist of this list
+        /// </summary>
+        public List<BasicBlock<N, D>> SuccessorEdges
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The Node of the program for which this control Flow Graph has been created.
+        /// </summary>
+        public N ProgramNode
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The Node of the procedure for which this control Flow Graph has been created.
+        /// </summary>
+        public N ProcedureNode
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// Initialize the construction of the Control Flow Graph.
+        /// </summary>
+        internal virtual void Initialize()
+        {
+            BlockFor = new Dictionary<N, BasicBlock<N, D>>();
+            AllBlocks = new List<BasicBlock<N, D>>();
+            RootBlocks = new List<BasicBlock<N, D>>();
+            SuccessorEdges = new List<BasicBlock<N, D>>();
+        }
+
+        /// <summary>
+        /// DFS Depth First Search implementation
+        /// </summary>
+        /// <param name="block">The current block</param>
+        /// <param name="discovered">Array of already discovered nodes</param>
+        /// <param name="callback">CallBack function</param>
+        internal void DFS(BasicBlock<N, D> block, System.Collections.BitArray discovered, BasicBlockCallback callback)
+        {
+            discovered[block.Index] = true;
+            if (!callback(block, this))
+                return;//Means stop
+            foreach (var edge in block.SuccessorEdges)
+            {
+                if (!discovered[SuccessorEdges[edge].Index])
+                {
+                    DFS(SuccessorEdges[edge], discovered, callback);
+                }
+            }
+        }
+
+        /// <summary>
+        /// DFS Depth First Search implementation
+        /// </summary>
+        /// <param name="rootBlock">The root block.</param>
+        /// <param name="callback">CallBack function</param>
+        public void DFS(BasicBlock<N, D> rootBlock, BasicBlockCallback callback)
+        {
+            System.Collections.BitArray discovered = new System.Collections.BitArray(AllBlocks.Count);
+            DFS(rootBlock, discovered, callback);
+        }
+
+        /// <summary>
+        /// DFS Depth First Search implementation.
+        /// </summary>
+        /// <param name="callback">CallBack function</param>
+        public void DFS(BasicBlockCallback callback)
+        {
+            foreach(var root in RootBlocks)
+            { 
+                DFS(root, callback);
```

Comment:
Call the most precise overload to avoid multiple BitArray instanciation.

Reply:
I don't understand what you mean.

Reply:
Here we are calling inside a `foreach` the overload of the method which instantiate the BitArray `discovered`. The array has the size of Allblocks which means it does not depend on where the traversal start, it accomodates with the whole graph no matter what. So to avoid creating it multiple times, I suggested to call `DFS(root, discovered, callback)` and create the discovered array outside of the foreach. This is not required though if we switch to a single RootBlock, this method will simply disappear.

Reply:
As discussed together, the array still needs to be instantiated because we need to reset the discovered nodes each time we start a new traversal from a specific root block.